compiler/rustc_metadata/src/locator.rs

```diff
@@ -783,14 +818,69 @@ fn get_metadata_section<'p>(
     filename: &'p Path,
     loader: &dyn MetadataLoader,
     cfg_version: &'static str,
+    dyn_crate: Option<Symbol>,
 ) -> Result<MetadataBlob, MetadataError<'p>> {
     if !filename.exists() {
         return Err(MetadataError::NotPresent(filename));
     }
+
+    if dyn_crate.is_some() && flavor != CrateFlavor::SDylib {
+        return Err(MetadataError::LoadFailure(
+            "`extern dyn` annotation is only avaible for dynamic dependencies with stable interface".to_string(),
+        ));
+    }
+
     let raw_bytes = match flavor {
         CrateFlavor::Rlib => {
             loader.get_rlib_metadata(target, filename).map_err(MetadataError::LoadFailure)?
         }
+        CrateFlavor::SDylib => {
+            let compiler = std::env::current_exe().map_err(|_err| {
+                MetadataError::LoadFailure(
+                    "couldn't obtain current compiler binary when loading `extern dyn` dependency"
+                        .to_string(),
+                )
+            })?;
+
+            let tmp_path = match tempfile::tempdir() {
+                Ok(tmp_path) => tmp_path,
+                Err(error) => {
+                    return Err(MetadataError::LoadFailure(format!(
+                        "Cannot create temporary directory: {}",
+                        error
+                    )));
+                }
+            };
+
+            let crate_name = dyn_crate.unwrap_or(kw::Empty);
```

Comment:
```suggestion
            let crate_name = dyn_crate.unwrap();
```
See the comment about assert above.